# 作业5 空间网络构建和查询

**作业目的：** 熟悉空间网络的基本概念和常见的空间网络分析和查询，掌握SQL3中的WITH RECURSIVE语句，熟悉从几何对象模型到空间网络模型的转换，掌握基于pgRouting的空间网络模型构建，熟悉pgRouting的最短路径算法，并会灵活运用解决一些空间网络的连通性查询问题，熟悉视图的可更新标准，掌握使用instead of触发器实现视图更新。

**注意事项：**
* SQL语句的错误输出为乱码时，修改SET client_encoding = 'GBK';或SET client_encoding = 'UTF-8';，重新连接数据库
* Jupyter Notebook对SQL语句的错误提示较弱，可以先在pgAdmin 4上执行，查看详细的错误信息
* 作业5总分60分，作业考察的题目后面标了具体分数，可以相互讨论思路，作业抄袭或雷同都要扣分
* **作业5\_学号\_姓名.ipynb**替换其中的学号和姓名，包含执行结果，和jsonData目录一起压缩为__作业5\_学号\_姓名.rar/zip__，**不要包含数据文件**，提交到学在浙大，作业5截止日期**2024.12.15**

### 0. With Recursive和pgRouting 3.4帮助文档

With Recursive是SQL3新增加的计算传递闭包语句，PostgreSQL实现了[With Recursive](http://www.postgresql.org/docs/current/static/queries-with.html)语句，请阅读并理解PostgreSQL帮助文档7.8.2 Recursive Queries的Recursive Query Evaluation步骤。**With Recursive语句需要避免死循环，如果运行时间过长，可以先在pgAdmin 4测试运行时间，或使用limit限制结果数目。**

pgRouting扩展了PostgreSQL/PostGIS地理空间数据库，提供了地理导航和网络分析功能。从几何对象模型构建空间网络模型，需要使用pgr_createTopology，pgr_analyzeGraph，pgr_nodeNetwork，地理导航可以使用pgr_dijkstra等最短路径算法。作业使用pgRouting 3.3及以上版本，请在使用相关函数前，仔细阅读[pgRouting](https://docs.pgrouting.org/latest/en/index.html)的函数帮助文档。

### 1. 观看访谈视频和阅读相关材料，回答问题（3分）

观看From GPS and Google Maps to Spatial Computing课程的[Module 3](http://www.cad.zju.edu.cn/home/ybtao/sdb/resources/Module%203.rar) Spatial Networks</a>的3-10 Dr. Dev Oliver at ESRI和3-11 Dr. Betsy George at Oracle Spatial的访谈视频（或阅读字幕），回答以下问题。

1.1  Dr. Betsy George提到飓风来临时，撤退方案不能直接使用最短路径算法，原因是什么？（1分）

1.2 Dr. Dev Oliver和Dr. Betsy George都谈到了企业使用的空间网络和课本上的空间网络的差异，至少给出2条差异描述。（2分）

### 2. 美国航空网络构建与查询（12分）

通过pgAdmin 4在PostgreSQL数据库中创建hw5数据库，增加postgis和pgRouting扩展(create extension postgis, create extension pgrouting)，并连接该数据库。
<img src = "usairports.png" width = 800>

In [11]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [12]:
%%sql postgresql://postgres:postgres@localhost:5432/hw5

SET statement_timeout = 0;
SET lock_timeout = 0;
SET client_encoding = 'UTF-8';
SET standard_conforming_strings = on;
SET check_function_bodies = false;
SET client_min_messages = warning;

Done.
Done.
Done.
Done.
Done.
Done.


[]

仔细阅读以下SQL语句，创建美国机场、机场关系和机场航班关系，理解航空网络的构建，并导入相应数据，完成以下查询。

In [5]:
%%sql

Drop Table if exists AIRPORT_LIST;
Drop Table if exists AIRPORT_NODE;
Drop Table if exists AIRPORT_LINK;

create table AIRPORT_LIST
(
    AIRPORT_ID   INT,
    AIRPORT_NAME VARCHAR(50)
);

create table AIRPORT_NODE
(
     NODE_ID      INT PRIMARY KEY,
     NODE_NAME    VARCHAR(200),
     NODE_TYPE    VARCHAR(200),
     ACTIVE       VARCHAR(1),
     PARTITION_ID INT,
     GEOMETRY     geometry(POINT, 4326)
);

create table AIRPORT_LINK
(
    LINK_ID         INT PRIMARY KEY,
    LINK_NAME       VARCHAR(200),
    START_NODE_ID   INT NOT NULL,
    END_NODE_ID     INT NOT NULL,
    LINK_TYPE       VARCHAR(200),
    ACTIVE          VARCHAR(1),
    LINK_LEVEL      INT,
    GEOMETRY        geometry(MultiLineString, 4326),
    COST            INT,
    BIDIRECTED      VARCHAR(1)                                    
);

copy airport_list from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\airport_list.txt' delimiter '#';
copy airport_node from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\airport_node.txt' delimiter '#';
copy airport_link from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\airport_link.txt' delimiter '#';

 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.
Done.
Done.
Done.
Done.
293 rows affected.
293 rows affected.
4093 rows affected.


[]

2.1 (练习题) 查询美国机场之间的单向航班数目，即存在从A到B的航班，但不存在从B到A的航班。

In [6]:
%%sql 
select l.link_id,l.start_node_id,l.end_node_id,l.cost
from airport_link l
except
select distinct l1.link_id,l1.start_node_id,l1.end_node_id,l1.cost
from airport_link l1, airport_link l2
where l1.start_node_id = l2.end_node_id
and l1.end_node_id = l2.start_node_id;


 * postgresql://postgres:***@localhost:5432/hw5
25 rows affected.


link_id,start_node_id,end_node_id,cost
3253,11292,13367,752
2403,14843,11042,1839
100,14709,10754,204
1729,10529,14492,532
956,11630,14709,373
1944,13230,11042,280
588,14730,10693,151
214,10732,11618,1585
967,11996,11042,450
3763,13891,10800,45


2.2 查询"Durango, CO"机场最多一次转机能够达到的机场名称和AIRPORT_ID，使用With Recursive实现。（2分）

非With Recursive实现

In [7]:
%%sql 
select end_node_id, (select airport_name from airport_list where airport_id = end_node_id)
from airport_list, airport_link
where airport_name = 'Durango, CO' and airport_id = start_node_id
union
select L2.end_node_id, (select airport_name from airport_list where airport_id = L2.end_node_id)
from airport_list, airport_link L1, airport_link L2
where airport_name = 'Durango, CO' and airport_id = L1.start_node_id and L1.end_node_id = L2.start_node_id

 * postgresql://postgres:***@localhost:5432/hw5
186 rows affected.


end_node_id,airport_name
12889,"Las Vegas, NV"
14689,"Santa Barbara, CA"
11109,"Colorado Springs, CO"
14193,"Pensacola, FL"
14783,"Springfield, MO"
10800,"Burbank, CA"
13303,"Miami, FL"
11057,"Charlotte, NC"
11503,"Eagle, CO"
11630,"Fairbanks, AK"


With Recursive实现，自行对比两种实现方式的优缺点

In [8]:
%%sql 
with recursive reach_airport as (
	select k.end_node_id, 1 as depth
	from airport_link k, airport_list t
	where k.start_node_id = t.airport_id and t.airport_name = 'Durango, CO'
	union all
	select k.end_node_id, r.depth+1
	from airport_link k, airport_list t
	inner join reach_airport r on r.end_node_id = t.airport_id
	where k.start_node_id = t.airport_id and r.depth<2
)
select distinct r.end_node_id, t.airport_name
from reach_airport r, airport_list t
where r.end_node_id = t.airport_id

 * postgresql://postgres:***@localhost:5432/hw5
186 rows affected.


end_node_id,airport_name
10721,"Boston, MA"
14457,"Rapid City, SD"
11986,"Grand Rapids, MI"
10713,"Boise, ID"
10821,"Baltimore, MD"
14307,"Providence, RI"
15376,"Tucson, AZ"
11540,"El Paso, TX"
12951,"Lafayette, LA"
10693,"Nashville, TN"


2.3 查询哪些机场最多一次转机能够达到"Bethel, AK"机场的机场名称和AIRPORT_ID，使用With Recursive实现。（2分）

In [9]:
%%sql
with recursive depart_airport as (
select k.start_node_id, 1 as depth
from airport_link k, airport_list t
where k.end_node_id = t.airport_id and t.airport_name = 'Bethel, AK'
union all
select k.start_node_id, d.depth+1
from airport_link k, airport_list t
inner join depart_airport d
on d.start_node_id = t.airport_id
where k.end_node_id = t.airport_id and d.depth<2
)
select distinct start_node_id, airport_name
from depart_airport d, airport_list t
where d.start_node_id = t.airport_id

 * postgresql://postgres:***@localhost:5432/hw5
26 rows affected.


start_node_id,airport_name
14771,"San Francisco, CA"
12266,"Houston, TX"
12173,"Honolulu, HI"
11298,"Dallas/Fort Worth, TX"
12954,"Long Beach, CA"
14869,"Salt Lake City, UT"
10245,"King Salmon, AK"
10170,"Kodiak, AK"
11630,"Fairbanks, AK"
13930,"Chicago, IL"


2.4 查询从"Dillingham, AK"机场到达 "Gainesville, FL"机场所需的最少转机次数，使用With Recursive实现。如果无法直接写出该语句，或运行时间过长，可以尝试枚举遍历法，即k从1，2，….不断增加，当k为多少，存在这样的路径。（3分）

In [10]:
%%sql 
with recursive depart_airport as (
select k.start_node_id, 1 as depth
from airport_link k, airport_list t
where k.end_node_id = t.airport_id and t.airport_name = 'Gainesville, FL'
union all
select k.start_node_id, d.depth+1
from airport_link k, airport_list t
inner join depart_airport d
on d.start_node_id = t.airport_id
where k.end_node_id = t.airport_id and d.depth<4
)
select distinct start_node_id, airport_name
from depart_airport d, airport_list t
where d.start_node_id = t.airport_id and airport_name = 'Dillingham, AK'

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.


start_node_id,airport_name
11336,"Dillingham, AK"


2.5 查询是否存在两个机场，无论经多少次转机都无法达到（即网络不连通）？使用With Recursive实现。（3分）

In [11]:
%%sql 
select * from (
with recursive f as(
	select 10135 as start_node_id
	union
	select lk.start_node_id
	from airport_link lk
	inner join f on lk.end_node_id=f.start_node_id
)
select airport_id from airport_list
except
select * from f
)
union
(
with recursive f as(
	select 10135 as end_node_id
	union
	select lk.end_node_id
	from airport_link lk
	inner join f on lk.start_node_id=f.end_node_id
)
select airport_id from airport_list
except
select * from f
)

 * postgresql://postgres:***@localhost:5432/hw5
0 rows affected.


airport_id


2.6 使用pgRouting的[pgr_dijkstra](http://docs.pgrouting.org/latest/en/pgr_dijkstra.html)函数，查询从"Dillingham, AK"机场到达 "Gainesville, FL"机场的最少花费路径，返回seq, node, edge, cost。（2分）

In [17]:
%%sql 
select seq, node, edge, cost from pgr_dijkstra(
	'SELECT link_id as id, start_node_id AS source, end_node_id AS target, cost FROM airport_link',
    (select airport_id from airport_list where airport_name='Dillingham, AK'),
    (select airport_id from airport_list where airport_name='Gainesville, FL'),
    true
)

 * postgresql://postgres:***@localhost:5432/hw5
5 rows affected.


seq,node,edge,cost
1,11336,742,329.0
2,10299,834,2519.0
3,13487,3499,907.0
4,10397,19,300.0
5,11953,-1,0.0


2.7 (练习题) pgRouting同时提供了[pgr_aStar](http://docs.pgrouting.org/latest/en/pgr_aStar.html)函数，对比Dijkstra算法与Astar算法，并说明在2.6题中为何不能使用pgr_aStar函数？Jupyter Notebook可以使用%%time给出cell的代码运行一次所花费的时间。

### 3. 杭州地铁网络构建与查询（9分）
地铁是典型的空间网络，杭州的12条运行的地铁线路，分别为：

1号线(杭州地铁1号线) 湘湖-滨康路-西兴-滨和路-江陵路-近江-婺江路-城站-定安路-龙翔桥-凤起路-武林广场-西湖文化广场-打铁关-闸弄口-火车东站-彭埠-七堡-九和路-九堡-客运中心-下沙西-金沙湖-高沙路-文泽路-文海南路-云水-下沙江滨-杭州大会展中心-港城大道-南阳-向阳路-萧山国际机场

2号线(杭州地铁2号线) 朝阳-曹家桥-潘水-人民路-杭发厂-人民广场-建设一路-建设三路-振宁路-飞虹路-盈丰路-钱江世纪城-钱江路-庆春广场-庆菱路-建国北路-中河北路-凤起路-武林门-沈塘桥-下宁桥-学院路-古翠路-丰潭路-文新-三坝-虾龙圩-三墩-墩祥街-金家渡-白洋-杜甫村-良渚

3号线(杭州地铁3号线)
（石马-星桥） 石马-小和山-屏峰-留下-西溪湿地南-花坞-东岳-古墩路-古荡新村-古荡-黄龙体育中心-黄龙洞-武林门-武林广场-西湖文化广场-潮王路-香积寺-大关-善贤-新天地街-汽轮广场-华丰路-同协路-桃花湖公园-丁桥-华鹤街-黄鹤山-星桥

（吴山前村-星桥) 吴山前村-汤家村(暂缓开通)-火车西站-龙舟北路-文一西路-绿汀路-创明路-全丰-高教路-联胜路-洪园-西溪湿地南-花坞-东岳-古墩路-古荡新村-古荡-黄龙体育中心-黄龙洞-武林门-武林广场-西湖文化广场-潮王路-香积寺-大关-善贤-新天地街-汽轮广场-华丰路-同协路-桃花湖公园-丁桥-华鹤街-黄鹤山-星桥

4号线(杭州地铁4号线) 池华街-金家渡-好运街-杭行路-储运路-平安桥-独城生态公园-吴家角港-桃源街-皋亭坝-新天地街-华中南路-笕桥老街-黎明-明石路-彭埠-火车东站-新风-新塘-景芳-钱江路-江锦路-市民中心-城星路-近江-甬江路-南星桥-复兴路-水澄桥-联庄-中医药大学-杨家墩-浦沿

5号线(杭州地铁5号线) 金星-绿汀路-葛巷-创景路-良睦路-杭师大仓前-永福路-五常-蒋村-浙大紫金港-三坝-萍水街-和睦-大运河-拱宸桥东-善贤-西文街-东园-杭氧-打铁关-宝善桥-建国北路-万安桥-城站-江城路-候潮门-南星桥-长河-聚才路-江晖路-滨康路-博奥路-金鸡路-人民广场-育才北路通惠中路-火车南站-双桥-姑娘桥

6号线(杭州地铁6号线)
(桂花西路-枸桔弄) 桂花西路-公望街-阳陂湖-高桥-富阳客运中心-受降-虎啸杏-银湖-野生动物园东-中村-音乐学院-美院象山-枫桦西路-之浦路-西浦路-中医药大学-伟业路-诚业路-建业路-长河-江汉路-江陵路-星民-奥体中心-博览中心-钱江世纪城-丰北-亚运村-三堡-昙花庵路-元宝塘-火车东站(东广场)-枸桔弄

(双浦-枸桔弄) 双浦-科海路-霞鸣街-美院象山-枫桦西路-之浦路-西浦路-中医药大学-伟业路-诚业路-建业路-长河-江汉路-江陵路-星民-奥体中心-博览中心-钱江世纪城-丰北-亚运村-三堡-昙花庵路-元宝塘-火车东站(东广场)-枸桔弄

7号线(杭州地铁7号线) 吴山广场-江城路-莫邪塘-观音塘-市民中心-奥体中心-兴议-明星路-建设三路-新兴路-新汉路-新街-合欢路-盈中-坎山-新港-萧山国际机场-永盛路-新镇路-义蓬-塘新线-青六中路-启成路-江东二路

8号线(杭州地铁8号线) 文海南路-工商大学云滨-桥头堡-河庄路-青西三路-青六中路-仓北村-冯娄村-新湾路

9号线(杭州地铁9号线) 观音塘-新业路-钱江路-江河汇-三堡-御道-五堡-六堡-红普南路-九睦路-客运中心-乔司南-乔司-翁梅-临平南高铁站-南苑-临平-邱山大街-何禹路-五洲路-龙安

10号线(杭州地铁10号线) 黄龙体育中心-文三路-学院路-翠柏路-北大桥-和睦-花园岗-渡驾桥-祥园路-杭行路-金德路-逸盛路

16号线(杭州地铁16号线) 九州街-临安广场-农林大学-青山湖-八百里-青山湖科技城-南峰-南湖-中泰-禹航路-凤新路-绿汀路

19号线(杭州地铁19号线) 苕溪-火车西站-创景路-海创园-西溪湿地北-文三路-沈塘桥-西湖文化广场-火车东站-御道-平澜路-耕文路-知行路-萧山国际机场-永盛路

百度地图实现了杭州地铁网络站点间的查询，[杭州轨道交通查询](https://map.baidu.com/subway/%E6%9D%AD%E5%B7%9E/@13376484.03,3517857.39,12z/ccode%3D179%26cname%3D%25E6%259D%25AD%25E5%25B7%259E)。

创建Line(id, lineName, name, geom(MultiLineString, 4326))，Station(id, name, geom(Point, 4326))，Link(id serial, fromStation, toStation, lineID)关系，根据提供的数据确定属性的数据类型，指定关系的主键和外键，并将lines_hangzhou.txt，stations_hangzhou.txt和links_hangzhou.txt导入相应关系中。

若由于编码方式不同导致无法数据导入，可以将文件另保存为其他编码方式再导入，或者修改set client_encoding = 'utf-8';。

In [13]:
from geom_display import display

In [18]:
%%sql
Drop Table if exists LINK cascade;
Drop Table if exists STATION cascade;
Drop Table if exists LINE cascade;

create table LINE
(
    id       INT PRIMARY KEY,
    lineName VARCHAR(50),
    name     VARCHAR(50),
    geom     geometry(MultiLineString, 4326)
);

create table STATION
(
     id   INT PRIMARY KEY,
     name VARCHAR(50),
     geom geometry(Point, 4326)
);

create table LINK
(
    id          serial PRIMARY KEY,
    fromStation INT,
    toStation   INT,
    lineID      INT,
    foreign key (fromStation) references STATION(id),
    foreign key (toStation)   references STATION(id),
    foreign key (lineID)      references LINE(id)
);

set client_encoding = 'utf-8';
copy line    from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\hangzhou_lines.txt'    delimiter '#';
copy station from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\hangzhou_stations.txt' delimiter '#';
copy link    from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\hangzhou_links.txt'    delimiter '#';

 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.
Done.
Done.
Done.
Done.
Done.
12 rows affected.
262 rows affected.
612 rows affected.


[]

实现以下地铁空间网络的分析与查询，注意不能修改上述关系，如增加属性，不能使用pgRouting函数实现。

3.1 给定地铁线路名称，如“杭州地铁1号线”，查询该线上的站点数目。(Find the number of stops on the Yellow West (YW) route)

In [19]:
%%sql 
select count(distinct fromstation) from link where lineid=1

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.


count
33


3.2 给定地铁站名称，如“浙大紫金港”，查询该地铁站所能到达的所有地铁站(可换乘, **路径长度不会超过50**)，使用With Recursive实现。(List all stops which can be reached from Downtown Berkeley (2))（2分）

In [20]:
%%sql 
with recursive rs as (
	select l.tostation as station_id,1 as depth
	from link l join station s
	on l.fromstation = s.id where s.name='浙大紫金港'
	union
	select l.tostation, rs.depth+1
	from link l join rs
	on l.fromstation = rs.station_id where rs.depth<50
)
select distinct station_id,station.name
from rs join station on rs.station_id = station.id

 * postgresql://postgres:***@localhost:5432/hw5
259 rows affected.


station_id,name
178,良睦路
206,野生动物园东
12,皋亭坝
35,下沙江滨
3,火车东站
226,新汉路
71,建业路
239,文三路
252,青西三路
154,翠柏路


3.3 给定两个地铁站名称，如“浙大紫金港”和“黄龙体育中心”，查询连接给定地铁站的路径，该路径经过的站点数目最少(假设地铁在任意两站之间的行驶时间相等)，即时间最短的路径(较快捷, **路径长度不会超过50**)，具体可查看[杭州轨道交通查询](https://map.baidu.com/subway/%E6%9D%AD%E5%B7%9E/@13376484.03,3517857.39,12z/ccode%3D179%26cname%3D%25E6%259D%25AD%25E5%25B7%259E)。上的路径查询效果，使用With Recursive实现。(List the routes numbers that connect Downtown Berkeley (2) & Daly City (5))（3分）

In [26]:
#查询经过站点数目最少的地铁路径。返回结果为三元组(gid站点id, name站点名称, geom站点位置),即路径上经过的所有站点
#若query1内容包含汉字，请用decode方法按照utf-8进行解码
query1 = """
with recursive bfs(from_id, to_id, depth, path) as (
	select l.fromstation as from_id, l.tostation as to_id, 1 as depth, array[l.fromstation, l.tostation] as path
	from link l join station s on l.fromstation = s.id
	where s.name = '浙大紫金港'
	union all
	select l.fromstation as from_id, l.tostation as to_id, bfs.depth+1, bfs.path || l.tostation
	from link l join bfs
	on l.fromstation = bfs.to_id
	where
	not (l.tostation = any(bfs.path))
    and bfs.depth < 20          
    and bfs.to_id <> (
		select s.id from station s where s.name = '黄龙体育中心'
	)
),
min_path as (
select *
from bfs
where bfs.to_id = (select s.id from station s where s.name = '黄龙体育中心')
order by depth
limit 1
)
select 
    s.id as gid,        
    s.name as name, 
    s.geom                  
from min_path mp,
     unnest(mp.path) with ordinality as  u(station_id, ord)
join station s on s.id = u.station_id
order by u.ord;
"""

result1 = %sql $query1

#根据results的路径查询结果，返回经过的地铁路径。返回结果为三元组(gid地铁线号, name其由linename和name拼接而成, geom地铁路线的几何信息)
query2 = """
with recursive bfs(from_id, to_id, depth, path) as (
	select l.fromstation as from_id, l.tostation as to_id, 1 as depth, array[l.fromstation, l.tostation] as path
	from link l join station s on l.fromstation = s.id
	where s.name = '浙大紫金港'
	union all
	select l.fromstation as from_id, l.tostation as to_id, bfs.depth+1, bfs.path || l.tostation
	from link l join bfs
	on l.fromstation = bfs.to_id
	where
	not (l.tostation = any(bfs.path))
    and bfs.depth < 20          
    and bfs.to_id <> (
		select s.id from station s where s.name = '黄龙体育中心'
	)
),
min_path as (
select *
from bfs
where bfs.to_id = (select s.id from station s where s.name = '黄龙体育中心')
order by depth
limit 1
),
result1 as (
select
	u.ord as rid,
    s.id as id,        
    s.name as name, 
    s.geom                  
from min_path mp,
     unnest(mp.path) with ordinality as  u(station_id, ord)
join station s on s.id = u.station_id
order by u.ord
)
select distinct l.lineid as gid, line.linename||line.name as name, line.geom
from result1 r1 join result1 r2 on r2.rid = r1.rid+1
join link l on l.fromstation = r1.id and l.tostation = r2.id
join line on l.lineid=line.id
"""
result2 = %sql $query2

display([result1, result2], "map1", 12, showToolTipLayer = 1, baseMapType = 0)

 * postgresql://postgres:***@localhost:5432/hw5
8 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
3 rows affected.


3.4 给定两个地铁站名称，如“浙大紫金港”和“古荡”，查询连接给定地铁站的路径，该路径换乘次数最少(**路径长度不会超过50**)，具体可查看[杭州轨道交通查询](https://map.baidu.com/subway/%E6%9D%AD%E5%B7%9E/@13376484.03,3517857.39,12z/ccode%3D179%26cname%3D%25E6%259D%25AD%25E5%25B7%259E)上的路径查询效果，使用With Recursive实现。(List the routes numbers that connect Downtown Berkeley (2) & Daly City (5))（4分）

In [29]:
#查询换乘次数最少的地铁路径。返回结果为三元组(gid站点id, name站点名称, geom站点位置),即路径上经过的所有站点
#若query1内容包含汉字，请用decode方法按照utf-8进行解码
query1 = """
with recursive lt(
	fromstation,
	tostation,
	lineid,
	trans,
	depth,
	path_station,
	path_line
) as (
	select l.fromstation, l.tostation, l.lineid, 0 as trans, 1 as depth, 
	array[l.fromstation, l.tostation] as path_station,
	array[l.lineid] as path_line
	from link l join station s on l.fromstation = s.id and s.name='浙大紫金港'
	union all 
	select l.fromstation, l.tostation, l.lineid, 
	case
		when l.lineid <> lt.lineid then lt.trans+1
		else lt.trans
	end as trans,lt.depth+1,lt.path_station || l.tostation,lt.path_line || l.lineid
	from link l join lt on l.fromstation = lt.tostation
	where not(l.tostation = any(lt.path_station)) and
	lt.depth<20 and lt.trans<5
),
less_trans_path as (
select * from lt
where lt.tostation = (select s.id from station s where s.name='古荡')
order by lt.trans asc, lt.depth asc
limit 1
)
select st.station_id as gid,s.name,s.geom
from less_trans_path ltp, 
unnest(ltp.path_station) with ordinality as st(station_id,ord)
join station s on s.id=st.station_id
order by ord
"""

result1 = %sql $query1

#根据results的路径查询结果，返回经过的地铁路径。返回结果为三元组(gid地铁线号, name其由linename和name拼接而成, geom地铁路线的几何信息)
query2 = """
with recursive lt(
	fromstation,
	tostation,
	lineid,
	trans,
	depth,
	path_station,
	path_line
) as (
	select l.fromstation, l.tostation, l.lineid, 0 as trans, 1 as depth, 
	array[l.fromstation, l.tostation] as path_station,
	array[l.lineid] as path_line
	from link l join station s on l.fromstation = s.id and s.name='浙大紫金港'
	union all 
	select l.fromstation, l.tostation, l.lineid, 
	case
		when l.lineid <> lt.lineid then lt.trans+1
		else lt.trans
	end as trans,lt.depth+1,lt.path_station || l.tostation,lt.path_line || l.lineid
	from link l join lt on l.fromstation = lt.tostation
	where not(l.tostation = any(lt.path_station)) and
	lt.depth<20 and lt.trans<5
),
less_trans_path as (
select * from lt
where lt.tostation = (select s.id from station s where s.name='古荡')
order by lt.trans asc, lt.depth asc
limit 1
)
select distinct pl.line_id as gid, line.linename || line.name as name,line.geom
from less_trans_path ltp, 
unnest(ltp.path_line) with ordinality as pl(line_id,ord)
join line on line.id=pl.line_id

"""
result2 = %sql $query2

display([result1, result2], "map2", 12, showToolTipLayer = 1, baseMapType = 0)

 * postgresql://postgres:***@localhost:5432/hw5
16 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
2 rows affected.


3.5 (练习题) 给定地铁线路名称，如“杭州地铁2号线”，查询该线上的起点或终点的地铁站。(Find the last stop on the Blue West (BW) route)

In [31]:
%%sql
with e as (
	select *
	from link
	where lineid=(select id from line where name='杭州地铁2号线')
),
neighborhood as(
select fromstation as a, tostation as b
from e
union
select tostation as a, fromstation as b
from e
),
edge_points as (
select a as station_id
from neighborhood
group by a
having count(*)=1
)
select s.id,s.name
from edge_points ep
join station s on s.id=ep.station_id

 * postgresql://postgres:***@localhost:5432/hw5
2 rows affected.


id,name
102,朝阳
171,良渚


3.6 (练习题) [杭州轨道交通查询](https://map.baidu.com/subway/%E6%9D%AD%E5%B7%9E/@13376484.03,3517857.39,12z/ccode%3D179%26cname%3D%25E6%259D%25AD%25E5%25B7%259E)提供了“较快捷”和“少换乘”两种查询模式，是否存在两个地铁站点的“较快捷”和“少换乘”路径不同？如果有请至少举例一对这样的站点，并修改3.3和3.4的站点名称，显示不同路径，如果没有，请说明理由。

### 4. 杭州道路网路构建与最短路径查询（28分）

Geocoding类库[geopy](https://github.com/geopy/geopy)利用the OpenStreetMap Nominatim, ESRI ArcGIS, Google Geocoding API (V3), Baidu Maps, Bing Maps API, Yahoo! PlaceFinder, Yandex, IGN France, GeoNames, NaviData, OpenMapQuest, What3Words, OpenCage, SmartyStreets, geocoder.us, and GeocodeFarm geocoder services，通过地址可以获得经纬度，或者通过经纬度获得地址，可以用于解决Lecture 9 Location Based Services中的Location: Where am I?问题。

基于OpenStreetMap上的杭州道路数据，包括poi(point of interest 点数据)和road(道路数据)，构建杭州道路网络，实现杭州道路上的最短路径查询。所建立的杭州道路网络为无向网络，在导航的过程中无需考虑道路单行道及走向问题。几何展示使用display函数，其查询结果至少包含gid，name和geom属性。

#### 4.0 Geocoding

Python函数address2location，输入地址字符串，返回经纬度。由于作业使用OpenStreetMap的道路数据，geocoding使用的是OpenStreetMap的Nominatim类。同一user_agent访问次数过多，可能会导致访问超时，可以尝试修改user_agent为特殊字符串。

In [32]:
from geopy.geocoders import Nominatim

def address2location(address):
    geolocator = Nominatim(user_agent = "sdb-app-2024", timeout = 1000)
    location   = geolocator.geocode(address)
    return (location.latitude, location.longitude)

In [33]:
addresses = ["浙江大学紫金港校区", "浙江大学玉泉校区", "浙江大学西溪校区", "浙江大学华家池校区"]

for address in addresses:
    point = address2location(address)
    print(address + "经纬度是(" + str(point[0]) + ", " + str(point[1]) + ")")

浙江大学紫金港校区经纬度是(30.305078, 120.0753432)
浙江大学玉泉校区经纬度是(30.2665272, 120.1180432)
浙江大学西溪校区经纬度是(30.2768278, 120.1366399)
浙江大学华家池校区经纬度是(30.270785, 120.1910915)


如果geocoding出错，超时无法返回经纬度信息，可以直接使用如下经纬度
* 浙江大学紫金港校区经纬度是(30.305078, 120.07534322932361)
* 浙江大学玉泉校区经纬度是(30.2665272, 120.1180431818039)
* 浙江大学西溪校区经纬度是(30.2768278, 120.13663990955736)
* 浙江大学华家池校区经纬度是(30.270785, 120.19109147440903)

#### 4.1 数据类型转换（3分）

道路的几何类型可能为ST_MultiLineString，如美国高速公路，但pgRouting是基于ST_LineString几何类型（思考为何不能使用ST_MultiLineString类型）。使用with recursive语句，将road_multilinestring关系的MultiLineString转换为LineString，保存在road_linestring关系中，**不能硬编码MultiLineString中的LineString的数目**。其中，name字段命名规则为road_multilinestring的name字段与该LineString的序号的拼接，中间用'.'分开。例如，"A"公路的MultiLineString包含4条LineString，则依次命名为"A.1", "A.2", "A.3"和"A.4"。

In [37]:
%%sql 
drop table if exists road_multilinestring;
CREATE TABLE road_multilinestring (
    gid serial primary key, 
    name character varying(20),
    geom geometry(MultiLineString, 4326)
);

insert into road_multilinestring(name, geom) values ('A', ST_GeomFromText('MultiLineString((1 1, 2 2, 3 3),(4 5, 6 7, 8 9),(4 5, 6 7, 8 9),(6 5, 4 3, 2 1))', 4326));
insert into road_multilinestring(name, geom) values ('B', ST_GeomFromText('MultiLineString((1 1, 2 2, 3 3),(4 5, 6 7, 8 9))', 4326));

drop table if exists road_linestring;
CREATE TABLE road_linestring (
    gid serial primary key, 
    name character varying(20),
    geom geometry(LineString, 4326)
);


 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.
1 rows affected.
1 rows affected.
Done.
Done.


[]

In [38]:
%%sql
insert into road_linestring (name,geom)
with recursive split_line(gid,name,n,geom) as (
	select gid,name,1 as n,st_geometryn(geom,1) as geom
	from road_multilinestring r
	where st_geometryn(geom,1) is not null
	union all
	select s.gid,s.name,n+1,st_geometryn(r.geom,n+1) as geom
	from road_multilinestring r
	join split_line s on r.gid = s.gid
	where st_geometryn(r.geom,n+1) is not null
)
select name || '.'||n as name,geom
from split_line sl
order by name

 * postgresql://postgres:***@localhost:5432/hw5
6 rows affected.


[]

In [39]:
%sql select name, ST_AsText(geom) from road_linestring order by name

 * postgresql://postgres:***@localhost:5432/hw5
6 rows affected.


name,st_astext
A.1,"LINESTRING(1 1,2 2,3 3)"
A.2,"LINESTRING(4 5,6 7,8 9)"
A.3,"LINESTRING(4 5,6 7,8 9)"
A.4,"LINESTRING(6 5,4 3,2 1)"
B.1,"LINESTRING(1 1,2 2,3 3)"
B.2,"LINESTRING(4 5,6 7,8 9)"


#### 4.2 杭州道路网络构建（3分）

实现人造道路上的道路网络构建，在pgAdmin 4中执行pgr_createTopology，pgr_analyzeGraph，pgr_nodeNetwork函数，生成道路空间网络模型。利用sql语句将自动生成的边表和顶点表信息分别插入到edges(注意len字段的更新)和nodes中。edges表的name字段命名规则为road的name字段与分割后subid字段的拼接，中间用'.'分开，例如，"A"道路经过分割后的道路名称未"A.1"和"A.2"，相邻的节点需要合并。
<img src = "roads.png">

In [55]:
%%sql
drop table if exists roads;
drop table if exists edges;
drop table if exists nodes;

create table roads (
    id integer NOT NULL,
    name text,
    geom geometry(LineString,4326)
);

insert into roads values (1, 'A', ST_GeomFromText('LineString(0 0, 10 0)', 4326));
insert into roads values (2, 'B', ST_GeomFromText('LineString(0 0, 0 10)', 4326));
insert into roads values (3, 'C', ST_GeomFromText('LineString(0 10, 10 10)', 4326));
insert into roads values (4, 'D', ST_GeomFromText('LineString(10 0, 10 15)', 4326));
insert into roads values (5, 'E', ST_GeomFromText('LineString(0 5, 10 15)', 4326));
insert into roads values (6, 'F', ST_GeomFromText('LineString(5 0, 10 5)', 4326));
insert into roads values (7, 'G', ST_GeomFromText('LineString(0 10, 10 0)', 4326));
insert into roads values (8, 'H', ST_GeomFromText('LineString(5 10, 15 0)', 4326));

create table edges (
       id serial primary key,
       name text,
       source integer,
       target integer,
       geom geometry(LineString, 4326),
       len float);

create table nodes (
       id serial primary key,
       geom geometry(Point, 4326)
);

 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.
Done.
Done.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
Done.
Done.


[]

In [56]:
%sql select pgr_nodeNetwork('roads',0.0001,'id','geom','noded')

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.


pgr_nodenetwork
OK


In [57]:
%sql select pgr_createTopology('roads_noded',0.0001,'geom','id')

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.


pgr_createtopology
OK


In [58]:
%%sql
insert into edges
select n.id,r.name||'.'||sub_id as name, source,target,n.geom,ST_length(n.geom)
from roads_noded n
join roads r on n.old_id = r.id

 * postgresql://postgres:***@localhost:5432/hw5
23 rows affected.


[]

In [59]:
%sql select id, name, source, target, ST_AsText(geom), len from edges;

 * postgresql://postgres:***@localhost:5432/hw5
23 rows affected.


id,name,source,target,st_astext,len
1,G.1,1,2,"LINESTRING(7.5 2.5,10 0)",3.5355339059327378
2,E.2,3,4,"LINESTRING(0 5,2.5 7.5)",3.5355339059327378
3,E.2,5,6,"LINESTRING(5 10,10 15)",7.0710678118654755
4,F.2,1,7,"LINESTRING(7.5 2.5,10 5)",3.5355339059327378
5,G.2,8,4,"LINESTRING(0 10,2.5 7.5)",3.5355339059327378
6,H.1,5,7,"LINESTRING(5 10,10 5)",7.0710678118654755
7,A.2,9,2,"LINESTRING(5 0,10 0)",5.0
8,D.2,2,7,"LINESTRING(10 0,10 5)",5.0
9,D.2,7,6,"LINESTRING(10 5,10 15)",10.0
10,D.3,7,10,"LINESTRING(10 5,10 10)",5.0


In [60]:
%sql insert into nodes select id,the_geom as geom from roads_noded_vertices_pgr

 * postgresql://postgres:***@localhost:5432/hw5
12 rows affected.


[]

In [61]:
%sql select id, ST_AsText(geom) from nodes;

 * postgresql://postgres:***@localhost:5432/hw5
12 rows affected.


id,st_astext
1,POINT(7.5 2.5)
2,POINT(10 0)
3,POINT(0 5)
4,POINT(2.5 7.5)
5,POINT(5 10)
6,POINT(10 15)
7,POINT(10 5)
8,POINT(0 10)
9,POINT(5 0)
10,POINT(10 10)


由于pgRounting的网络构建的随机性和杭州道路的复杂性，杭州道路网路将直接导入，用于4.3-4.8的道路网络查询。

In [68]:
%%sql
set client_encoding = 'GBK';
drop table if exists poi; 
drop table if exists edges;
drop table if exists nodes;

create table poi (
    id   integer NOT NULL,
    lon  double precision,
    lat  double precision,
    name text,
    geom geometry(Point,4326)
);

create table edges (
       id     serial primary key,
       name   text,
       source int,
       target int,
       geom   geometry(LineString, 4326),
       len    float);

create table nodes (
       id   serial primary key,
       name text,
       geom geometry(Point, 4326)
);

copy poi   from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\poi.txt'   delimiter '#';
set client_encoding = 'utf-8';
copy edges from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\edges.txt' delimiter '#';
copy nodes from  'D:\\Code_File\\GeoSpatialDatabase\\hw5_data\\nodes.txt' delimiter '#';


 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.
Done.
Done.
Done.
Done.
Done.
2200 rows affected.
Done.
10673 rows affected.
6766 rows affected.


[]

#### 4.3 最近的道路网络节点（2分）

在路径导航过程中，假设出发和目的地都先走路到道路网络节点，通过ST_Location2Node函数获得当前位置最近的道路网络节点，再通过道路网络节点之间的最短距离实现Lecture 9 Location Based Services的Routes: How do I get there?问题。

实现ST_Location2Node函数，输入经纬度位置，输出道路网络中，离该位置直线距离最近的道路网络端点。

In [69]:
%%sql
create or replace function ST_Location2Node(lat float, lon float) 
    returns integer
as $$
declare num integer;
begin
    select id into num 
    from nodes
    order by geom <-> ST_SetSRID(ST_MakePoint(lon, lat), 4326)
    limit 1;
    
    return num;
end;
$$ language plpgsql;

 * postgresql://postgres:***@localhost:5432/hw5
Done.


[]

In [70]:
addresses = ["浙江大学紫金港校区", "浙江大学玉泉校区", "浙江大学西溪校区", "浙江大学华家池校区"]

for address in addresses:
    point = address2location(address)
    query = "select ST_Location2Node(%s, %s)" % (point[0], point[1])
    result = %sql $query
    print(address + "直线距离最近的网络节点是" + str(result[0][0]))

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
浙江大学紫金港校区直线距离最近的网络节点是422
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
浙江大学玉泉校区直线距离最近的网络节点是789
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
浙江大学西溪校区直线距离最近的网络节点是1044
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
浙江大学华家池校区直线距离最近的网络节点是2057


如果geocoding出错，超时无法返回经纬度信息，可以直接使用如下网络节点
* 浙江大学紫金港校区直线距离最近的网络节点是422
* 浙江大学玉泉校区直线距离最近的网络节点是789
* 浙江大学西溪校区直线距离最近的网络节点是1044
* 浙江大学华家池校区直线距离最近的网络节点是2057

#### 4.4 导航路径生成（Dijkstra算法）（2分）

根据4.3的查询结果，使用pg_dijkstra算法，查询从紫金港校区到西溪校区的最短驾驶距离对应的路线，查询结果至少包含gid，name和geom属性。

In [72]:
query = """    
select id as gid, e.name as name, e.geom as geom from pgr_dijkstra(
	'select id,source,target,len as cost from edges',
	422,
	1044,
	false
) d
join edges e on e.id=d.edge
"""
result = %sql $query

display([result], "map3", 13)

 * postgresql://postgres:***@localhost:5432/hw5
32 rows affected.


#### 4.5 驾驶距离最近的电影院（6分）

Lecture 9 Location Based Services的Directory: What is around me?，例如查询浙江大学西溪校区<b>直线距离最近</b>的电影院，注意本题共有**5处(修改此处)**需要修改。

In [74]:
point = address2location("浙江大学西溪校区")
query = """select id, name, ST_AsText(geom) from poi where name like '%%电影%%' or 
           name like '%%影院%%' or name like '%%影城%%' 
            order by geom <-> ST_SetSRID(ST_MakePoint(%s, %s), 4326)
            limit 1;"""% (point[1], point[0])
print(query)
result = %sql $query
print(result)

select id, name, ST_AsText(geom) from poi where name like '%电影%' or 
           name like '%影院%' or name like '%影城%' 
            order by geom <-> ST_SetSRID(ST_MakePoint(120.1366399, 30.2768278), 4326)
            limit 1;
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
+------+----------+----------------------------+
|  id  |   name   |         st_astext          |
+------+----------+----------------------------+
| 2199 | 黄龙影城 | POINT(120.141376 30.27236) |
+------+----------+----------------------------+


实际上，希望查询驾驶距离最近的电影院('%电影%' or '%影院%', or '%影城%')，而非直线距离最近。基于当前位置的经纬度，输出驾驶距离最近的电影院，POI的id。忽略走路距离，位置到最近网络节点编号查询使用ST_Location2Node函数。通过查询最短驾驶距离，也实现了Lecture 9 Location Based Services的Routes: How do I get there?问题。

实现思路：依次遍历所有电影院，通过Dijkstra最短路径算法获得路径，计算总的路程，获得驾驶距离最短的电影院编号。

思考是否有更高效的方法，减少最短路径查询次数。

In [75]:
point = address2location("浙江大学西溪校区")
result = %sql select id, ST_X(geom), ST_Y(geom) from poi where name like '%电影%' or name like '%影院%'  or name like '%影城%'
print(len(result))

 * postgresql://postgres:***@localhost:5432/hw5
5 rows affected.
5


In [83]:
point = address2location("浙江大学西溪校区")
result = %sql select id, ST_X(geom), ST_Y(geom) from poi where name like '%电影%' or name like '%影院%'  or name like '%影城%'
print(len(result))

cinmaID = -1
minLength = 1e10
for cinma in result:
    query1 = "select ST_Location2Node(%f, %f) as id" %(point[0], point[1])
    closestp1 = %sql $query1
    query2 = "select ST_Location2Node(%f, %f) as id" %(cinma[2], cinma[1])
    closestp2 = %sql $query2
    query = """
            select max(agg_cost) from pgr_dijkstra(
                'select id,source,target,len as cost from edges',
                %d,
                %d,
                false
            ) d
            join edges e on e.id=d.edge
            """ %(closestp1[0][0], closestp2[0][0])
    length = %sql $query
    if length[0][0] < minLength:
        minLength = length[0][0]
        cinmaID = cinma[0]

print(cinmaID, minLength)

 * postgresql://postgres:***@localhost:5432/hw5
5 rows affected.
5
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgre

在地图上展示从浙江大学西溪校区到其驾驶距离最近电影院的导航路径

In [84]:
point = address2location("浙江大学西溪校区")
query1 = "select ST_Location2Node(%f, %f) as id" %(point[0], point[1])
closestp1 = %sql $query1
print(closestp1[0][0])

cinma = %sql select ST_X(geom), ST_Y(geom) from poi where id = 2199
query2 = "select ST_Location2Node(%f, %f) as id" %(cinma[0][1], cinma[0][0])
closestp2 = %sql $query2
print(closestp2[0][0])

#查询这两个网络节点之间的最短路径，输出为(gid,name,geom)三元组
query1 = """
select id as gid, e.name as name, e.geom as geom from pgr_dijkstra(
	'select id,source,target,len as cost from edges',
	%d,
	%d,
	false
) d
join edges e on e.id=d.edge
""" %(closestp1[0][0], closestp2[0][0])
result1 = %sql $query1

#查询这两个网络节点的几何信息，输出为(gid,name,geom)三元组
query2 = """
select id as gid, e.name as name, e.geom as geom from pgr_dijkstra(
	'select id,source,target,len as cost from edges',
	%d,
	%d,
	false
) d
join edges e on e.id=d.edge
""" %(closestp1[0][0], closestp2[0][0])
result2 = %sql $query2

display([result1, result2], "map4", 13)

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
1044
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
3897
 * postgresql://postgres:***@localhost:5432/hw5
6 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
6 rows affected.


#### 4.6 导航路径推荐 （4分）

当查询从A到B的驾驶路线时，地图服务商（例如高德地图）通常会提供几条路线供用户选择，其中一条是最短驾驶距离对应的路线，其他路线可能会考虑当时的交通状况，例如某条道路当前比较拥堵，行驶缓慢，将提供避开此道路的最短驾驶距离对应的路线。

根据4.3的查询结果，基于Dijkstra算法生成从紫金港校区到玉泉校区的最短驾驶距离对应的路线，查询结果至少包含gid，name和geom属性。

<img src="routes.png" width = 400>

In [86]:
query = """    
select pgr.edge as gid, e.name, e.geom from pgr_dijkstra(
    'select id,source,target,len as cost from edges',
    422,
    789,
    false
) pgr
join edges e on pgr.edge=e.id
"""
result = %sql $query

display([result], "map5", 13)

 * postgresql://postgres:***@localhost:5432/hw5
27 rows affected.


假设以下道路存在拥堵

    "竞舟路.9.1.1.1.1.1.1", 起点为"POINT(120.0987188 30.2888717)", 终点为"POINT(120.1000613 30.2841765)" 
    "西溪路 Xixi Road.22.1.1.1.1.1.1", 起点为"POINT(120.1197977 30.2719367)", 终点为"POINT(120.1240717 30.2703345)"
    
根据4.3的查询结果，基于Dijkstra算法生成此时从紫金港校区到玉泉校区不包含上述道路的最短驾驶距离对应的路线，查询结果至少包含gid，name和geom属性。
  


In [93]:
query = """ 
select pgr.edge as gid, e.name, e.geom from pgr_dijkstra(
    $$select id,source,target,len as cost from edges where name not in ('竞舟路.9.1.1.1.1.1.1','西溪路 Xixi Road.22.1.1.1.1.1.1')$$,
    422,
    789,
    false
) pgr
join edges e on pgr.edge=e.id
"""
result = %sql $query

display([result], "map6", 13)

 * postgresql://postgres:***@localhost:5432/hw5
22 rows affected.


#### 4.7 导航偏离下重新导航（4分）

根据4.3的查询结果，基于Dijkstra算法生成从紫金港校区到西溪校区的最短驾驶距离对应的路线，查询结果至少包含gid，name和geom属性。

In [94]:
query = """
select pgr.edge as gid, e.name, e.geom from pgr_dijkstra(
    'select id,source,target,len as cost from edges',
    422,
    1044,
    false
) pgr
join edges e on pgr.edge=e.id
"""
result = %sql $query

display([result], "map7", 13)

 * postgresql://postgres:***@localhost:5432/hw5
32 rows affected.


当系统发现用户偏离了原始的导航路线，根据当前情况，将自动重新计算最短驾驶距离对应的路线。这里涉及到<a href="https://mapcoding.cn/map-maching/" target="_blank">道路匹配</a>模块，即将用户匹配到某条道路上。

假设在实际行驶过程中，某车本应从"文一西路.24.1.1.1.1.1.1"转向"竞舟路.9.1.1.1.1.1.1"，却前行到"文一西路.25.1.1.1.1.1.1"（中间有绿化带，不能随意掉头返回到"竞舟路.9.1.1.1.1.1.1"）。根据车所在的位置和行驶方向，基于Dijkstra算法重新计算最短驾驶距离对应的路线。

    "文一西路.24.1.1.1.1.1.1", 起点为"POINT(120.0940602 30.2887916)", 终点为"POINT(120.0987188 30.2888717)"
    "竞舟路.9.1.1.1.1.1.1", 起点为"POINT(120.0987188 30.2888717)", 终点为"POINT(120.1000613 30.2841765)"   
    "文一西路.25.1.1.1.1.1.1", 起点为"POINT(120.0987188 30.2888717)", 终点为"POINT(120.102833 30.2890362)" 
 

In [97]:

query = """
select pgr.edge as gid, e.name, e.geom from pgr_dijkstra(
    $$select id,source,target,len as cost from edges where name <> '竞舟路.9.1.1.1.1.1.1'$$,
    (select target from edges where name='文一西路.25.1.1.1.1.1.1'),
    1044,
    false
) pgr
join edges e on pgr.edge=e.id
"""
result = %sql $query

display([result], "map8", 15)

 * postgresql://postgres:***@localhost:5432/hw5
20 rows affected.


#### 4.8 红绿灯最少的路径（4分）

假设每个节点都有红绿灯，根据4.3的查询结果，查询从紫金港校区到西溪校区的经过红绿灯最少的路线，查询结果至少包含gid，name和geom属性。该路线忽略道路长度，仅考虑路线经过的红绿灯数目，不能修改杭州道路网络。

In [98]:
query = """
select pgr.edge as gid, e.name, e.geom from pgr_dijkstra(
    'select id,source,target,1 as cost from edges',
    422,
    1044,
    false
) pgr
join edges e on pgr.edge=e.id
"""
result = %sql $query

display([result], "map9", 13)

 * postgresql://postgres:***@localhost:5432/hw5
28 rows affected.


进一步要求经过红绿灯最少的路线长度不能超过最短路径长度的1.5倍，查询结果至少包含gid，name和geom属性，不能修改杭州道路网络。

In [14]:
query = """
with baseline as (
    select sum(ST_Length(e.geom)) as min_len
    from pgr_dijkstra(
        'SELECT id, source, target, ST_Length(geom) as cost, ST_Length(geom) as reverse_cost FROM edges',
        422, 
        1044,
        false
    ) d
    join edges e on d.edge = e.id
),
ksp_results as (
    select path_id,path_seq,edge, cost       
    from pgr_ksp(
        'SELECT id, source, target, 1 as cost, 1 as reverse_cost FROM edges',
        422, 
        1044,
        50,
        directed := false
    )
),
path_stats as (
    select 
        k.path_id,
        sum(ST_Length(e.geom)) as real_length, 
        sum(k.cost) as lights_count            
    from ksp_results k
    join edges e on k.edge = e.id
    group by k.path_id
),
best_path as (
    select 
        p.path_id
    from path_stats p, baseline b
    where p.real_length <= (b.min_len * 1.5)
    order by p.lights_count asc, p.real_length asc
    limit 1
)
select
    k.edge as gid,  
    e.name,         
    e.geom          
from best_path bp
join ksp_results k on bp.path_id = k.path_id
join edges e on k.edge = e.id
order by k.path_seq;
"""
result = %sql $query

display([result], "map10", 13)

 * postgresql://postgres:***@localhost:5432/hw5
28 rows affected.


### 5. 视图与触发器（8分）

track关系可以用于分析道路拥堵，即道路上车辆数目。所涉及的数据包括杭州道路数据edges(id, name, source, target, geom, len)和车辆位置数据track(time, position, userName, carID)。

In [99]:
%%sql
drop table if exists track cascade;

create table track(
    time timestamp default CURRENT_TIMESTAMP,
    position geometry(POINT, 4326),
    userName text default SESSION_USER,
    carID text
);

insert into track values('2024-11-28 10:20:08', ST_GeomFromText('point(120.104686575 30.283505885)',4326), 'Jack' , '101');
insert into track values('2024-11-28 10:20:12', ST_GeomFromText('point(120.10475310 30.28328588)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:13', ST_GeomFromText('point(120.104819625 30.283065875)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:15', ST_GeomFromText('point(120.10488615 30.28284587)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:18', ST_GeomFromText('point(120.104952675 30.282625865)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:23', ST_GeomFromText('point(120.104819625 30.283065875)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:25', ST_GeomFromText('point(120.104979285 30.282537863)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:28', ST_GeomFromText('point(120.1049992425 30.2824718615)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:31', ST_GeomFromText('point(120.10501920 30.28240586)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:33', ST_GeomFromText('point(120.105045810 30.282317858)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:35', ST_GeomFromText('point(120.105085725 30.282185855)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:39', ST_GeomFromText('point(120.105125640 30.282053852)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:43', ST_GeomFromText('point(120.105178860 30.281877848)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:47', ST_GeomFromText('point(120.105218775 30.281745845)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:52', ST_GeomFromText('point(120.105298605 30.281481839)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:59', ST_GeomFromText('point(120.105378435 30.281217833)',4326), 'Jack', '101');
insert into track values('2024-11-28 10:20:29', ST_GeomFromText('point(120.10475310 30.28328588)',4326), 'David', '102');
insert into track values('2024-11-28 10:20:34', ST_GeomFromText('point(120.1047810405 30.2831934779)',4326), 'David', '102');
insert into track values('2024-11-28 10:20:45', ST_GeomFromText('point(120.10475310 30.28328588)',4326), 'Tom', '103');
insert into track values('2024-11-28 10:20:53', ST_GeomFromText('point(120.1047770490 30.2832066782)',4326), 'Tom', '103');
insert into track values('2024-11-28 10:20:31', ST_GeomFromText('point(120.103880996283 30.2860733641809)',4326), 'Sally', '104');
insert into track values('2024-11-28 10:20:35', ST_GeomFromText('point(120.1039895370264 30.28572229134472)',4326), 'Sally', '104');

 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.


[]

5.1 首先创建视图CurrentTrack(carID, position, roadID)，表示每辆车**当前**所在位置及该位置所在的道路，即距离所在位置最近的道路。（2分）

In [7]:
%%sql
drop view if exists currenttrack;
create view CurrentTrack
    as
    select distinct on(carid) carid as carID,position,(
    select id from edges order by geom<->position limit 1
    ) as roadID
    from track
    order by carid,time desc

 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.


[]

5.2 基于CurrentTrack视图，构造SQL语句查询道路上车数目，查询返回道路编号和道路上车的数目，按车的数目降序排列。（2分）

In [9]:
%%sql
select roadID, count(*) as count from CurrentTrack 
join edges 
on CurrentTrack.roadID = edges.id
group by CurrentTrack.roadID
order by count desc

 * postgresql://postgres:***@localhost:5432/hw5
2 rows affected.


roadid,count
68364,3
68368,1


5.3 根据**SQL**的视图可更新标准，该视图是否为可更新视图，请说明理由。（1分）

5.4 根据**PostgreSQL数据库**的视图可更新标准，该视图是否为可更新视图，请说明理由。（1分）

5.5 为该视图创建触发器实现数据插入，roadID无需插入。（2分）

In [15]:
%%sql
create or replace function insert_currenttrack_func()
returns trigger as $$
begin
    insert into track(time,position,userName,carID)
    values(
        now(),
        new.position,
        (select username from (select distinct username, carid from track) where carid=new.carid),
        new.carID
    );
    return new;
end;
$$ language plpgsql;
create or replace trigger trg_insert_currenttrack
instead of insert
on currenttrack
for each row
execute function insert_currenttrack_func();

 * postgresql://postgres:***@localhost:5432/hw5
Done.
Done.


[]

In [16]:
%sql insert into CurrentTrack(carID, position) values('102', ST_GeomFromText('Point(120.10475310 30.28328588)', 4326));

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.


[]

In [17]:
%sql select time, ST_AsText(position), username from track where carID = '102' order by time desc

 * postgresql://postgres:***@localhost:5432/hw5
3 rows affected.


time,st_astext,username
2025-12-22 09:19:13.583444,POINT(120.1047531 30.28328588),David
2024-11-28 10:20:34,POINT(120.1047810405 30.2831934779),David
2024-11-28 10:20:29,POINT(120.1047531 30.28328588),David


### 6. 空间函数实现（附加题4分）

假设所有点都在2D平面上，即二维笛卡尔坐标系，不考虑球表面（<a href="http://postgis.net/docs/using_postgis_dbmanagement.html#PostGIS_Geography" target="_blank">geography类型</a>），所有的几何都是简单几何，请用PostgreSQL的<a href="http://www.postgresql.org/docs/current/static/plpgsql.html" target="_blank">PL/pgSQL语言</a>实现ST_GeomBoundary和ST_LineTouches函数。只可以使用PostgreSQL的<a href="http://www.postgresql.org/docs/current/static/functions-math.html" target="_blank">数学函数库</a>与PostGIS的ST_GeometryType、ST_NumGeometries、ST_GeometryN、ST_ExteriorRing、ST_NumInteriorRings、ST_InteriorRingN、ST_NumPoint、ST_PointN、ST_StartPoint、ST_EndPoint、ST_X、ST_Y、ST_Union、ST_Intersection、ST_Difference、ST_MakePoint、ST_MakeLine和ST_Collect函数。

#### 6.1 几何边界

几何边界函数 <a href="http://postgis.net/docs/ST_Boundary.html" target="_blank">ST_Boundary</a>

	geometry ST_GeomBoundary(geometry g)
    
要求实现Point、LineString、Polygon、MultiPoint、MultiLineString和MultiPolygon几何类型的边界获取，其他类型返回GEOMETRYCOLLECTION EMPTY。忽略MultiPoint(10 10, 20 20)和MultiPoint(20 20, 10 10)的差异，即与ST_GeomBoundary结果只要ST_Equals就满足题目要求。

In [8]:
%%sql
create or replace function st_geomboundary(geom geometry)
returns geometry as $$
declare
    g_type text;
    res geometry := null;
    sub_geom geometry;
    temp_geom geometry;
    p1 geometry;
    p2 geometry;
    pt geometry;
    i integer;
    j integer;
    num_geoms integer;
    num_rings integer;
    check_inter geometry;
begin
    if geom is null then
        return null;
    end if;
    
    g_type := st_geometrytype(geom);

    if g_type = 'ST_Point'  then
        return 'POINT EMPTY'::geometry;

    elsif g_type = 'ST_MultiPoint' then
        return 'MULTIPOINT EMPTY'::geometry;

    elsif g_type = 'ST_LineString' then
        p1 := st_startpoint(geom);
        p2 := st_endpoint(geom);
        
        if (st_x(p1) = st_x(p2) and st_y(p1) = st_y(p2)) then
            return 'MULTIPOINT EMPTY'::geometry;
        else
            return st_collect(p1, p2);
        end if;

    elsif g_type = 'ST_MultiLineString' then
        num_geoms := st_numgeometries(geom);
        
        for i in 1..num_geoms loop
            sub_geom := st_geometryn(geom, i);
            p1 := st_startpoint(sub_geom);
            p2 := st_endpoint(sub_geom);

            if (st_x(p1) = st_x(p2) and st_y(p1) = st_y(p2)) then
                continue;
            end if;

            for pt in select unnest(array[p1, p2]) loop
                if res is null then
                    res := pt;
                else
                    check_inter := st_intersection(res, pt);
                    
                    if st_numpoints(check_inter) > 0 then
                        res := st_difference(res, pt);
                        
                        if st_numpoints(res) = 0 then
                            res := null;
                        end if;
                    else
                        res := st_union(res, pt);
                    end if;
                end if;
            end loop;
        end loop;

        if res is null then
            return 'GEOMETRYCOLLECTION EMPTY'::geometry;
        else
            return res;
        end if;

    elsif g_type = 'ST_Polygon' then
        res := st_exteriorring(geom);
        num_rings := st_numinteriorrings(geom);
        
        if num_rings > 0 then
            for i in 1..num_rings loop
                res := st_collect(res, st_interiorringn(geom, i));
            end loop;
        end if;
        
        return res;

    elsif g_type = 'ST_MultiPolygon' then
        num_geoms := st_numgeometries(geom);
        
        for i in 1..num_geoms loop
            sub_geom := st_geometryn(geom, i);
            
            temp_geom := st_exteriorring(sub_geom);
            num_rings := st_numinteriorrings(sub_geom);
            if num_rings > 0 then
                for j in 1..num_rings loop
                    temp_geom := st_collect(temp_geom, st_interiorringn(sub_geom, j));
                end loop;
            end if;

            res := st_collect(res, temp_geom);
        end loop;
        
        return res;

    else
        return 'GEOMETRYCOLLECTION EMPTY'::geometry;
    end if;

end;
$$ language plpgsql;

 * postgresql://postgres:***@localhost:5432/hw5
Done.


[]

In [9]:
geometries = ["Point(10 50)", 
              "MultiPoint(10 50, 40 60)",
              "LineString(10 50, 20 30)",
              "LineString(10 50, 20 30, 10 50)", 
              "MultiLineString((10 50, 20 30, 10 30),(30 60, 50 60, 10 40))",
              "MultiLineString((10 50, 20 30, 10 30),(10 50, 50 60, 10 30))",
              "MultiLineString((10 50, 20 30, 10 30),(10 50, 50 60, 10 50))",
              "Polygon((10 50, 20 30, 40 60, 10 50),(10 50, 20 30, 40 60, 10 50),(10 50, 20 30, 40 60, 10 50))",
              "MultiPolygon(((10 50, 20 30, 40 60, 10 50),(10 50, 20 30, 40 60, 10 50)))"
             ]
template = "SELECT ST_AsText(ST_GeomBoundary('%s'::geometry)), ST_AsText(ST_Boundary('%s'::geometry))"

passedTests = len(geometries)
for geometry in geometries:
    query  = template % (geometry, geometry)
    result = %sql $query
    if result[0][0] != result[0][1]:
        passedTests -= 1
        print(geometry + ' boundary test failed\nYour result is ' + str(result[0][0]) + '\nPGIS result is ' + str(result[0][1]))

print(str(passedTests) + ' / ' + str(len(geometries)) + " tests are passed")

 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
MultiLineString((10 50, 20 30, 10 30),(30 60, 50 60, 10 40)) boundary test failed
Your result is MULTIPOINT((10 30),(10 50),(30 60),(10 40))
PGIS result is MULTIPOINT((10 50),(10 30),(30 60),(10 40))
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
MultiLineString((10 50, 20 30, 10 30),(10 50, 50 60, 10 30)) boundary test failed
Your result is MULTIPOINT((10 30),(10 50))
PGIS result is MULTIPOINT EMPTY
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
 * postgresql://postgres:***@localhost:5432/hw5
1 rows affected.
Polygon((10 50, 20 30, 40 60, 10 50),(10 50, 20 30, 40 60, 10 50),(10 50, 20 30, 40 60, 10 50)) boundary test failed
Your r

#### 6.2 折线与折线相接空间关系判断

几何相接空间关系判断函数 <a href="http://postgis.net/docs/ST_Touches.html" target="_blank">ST_Touches</a>

	boolean ST_LineTouches(geometry g1, geometry g2)
    
几何要素g1和g2类型为ST_LineString，判断线与线是否相接，当几何类型不符合要求时，返回GEOMETRYCOLLECTION EMPTY。

In [ ]:
%%sql
create or replace function ST_LineTouches(g1 geometry, g2 geometry)
        returns boolean


In [ ]:
linegroupA = [""]
linegroupB = [""]

template = "SELECT ST_LineTouches('%s'::geometry, '%s'::geometry), ST_Touches('%s'::geometry, '%s'::geometry)"

passedTests = len(linegroupA)
for i in range(len(linegroupA)):
    lineA  = linegroupA[i]
    lineB  = linegroupB[i]
    query  = template % (lineA, lineB, lineA, lineB)
    result = %sql $query
    if result[0][0] != result[0][1]:
        passedTests -= 1
        print(lineA + ' and ' + lineB + ' touch test failed\nYour result is ' + str(result[0][0]) + '\nPGIS result is ' + str(result[0][1]))

print(str(passedTests) + ' / ' + str(len(linegroupA)) + " tests are passed")

### 作业感想（6分）

这是地理空间数据库课程的最后一个个人作业，填写本次作业感想有加分哦<br/>

课程内容可以分成六大主题：关系数据库查询(RA+SQL), 几何对象模型与查询(Geometry+Spatial Analysis+PostGIS)，空间数据库设计(E/R+BCNF), 物理模型(Spatial Curve+Spatial Index+Query Plan), 空间网络模型与查询(With Recursive+pgRouting), 数据库安全(View+Trigger+Transactions)。

对课程内容有什么建议，比如(1) 哪些主题或内容较为简单或其他课程已有介绍，可以少讲或不讲，(2) 哪些主题或内容较难，不易理解，需要详细介绍或不讲在后续课程中学习（如空间索引实现、PostgreSQL服务器编程、pgRouting基于几何对象模型的网络模型构建等），(3) 希望增加哪些数据库相关的内容，这些内容其他课程没有介绍、或比较重要、或是未来发展趋势，(4) 其他关于课程内容的建议

课程作业包括5次学在浙大测试、5次个人作业（对应前五个主题）、1次课堂测试和1次小组作业，占课程成绩的50%。

对课程作业有什么建议，比如(1) 哪些作业的题目太简单或太难或与其他课程重复，是否能够帮助你理解课程内容，了解数据库的应用和价值，(2) 作业的难易程度，平均每次作业完成需要多少时间，希望提供哪些帮助，(3) 作业提供的类库是否有帮助，如geom_display和relation_algebra，希望提供哪些功能，(4) 作业的附加题是否有必要，希望提供哪些类型的附加题，（5）其他关于课程作业的建议

对课程的其他建议，比如课程提供的课件、练习、互动答题和阅读材料，有无必要或如何发挥更大的作用等